packages/zoltan2/core/src/algorithms/color/Zoltan2_AlgHybrid2GL.hpp

```diff
@@ -0,0 +1,1532 @@
+          if((per_proc_round_adj_sums[curr_round][proc_to_send] + recvDegrees[j])*sizeof(gno_t) > INT_MAX){
+            curr_round++;
+          }
+          per_proc_round_adj_sums[curr_round][proc_to_send] += recvDegrees[j];
+          per_proc_round_vtx_sums[curr_round][proc_to_send]++;
+        }
+      }
+      
+      if(verbose) std::cout<<comm->getRank()<<": building recv GID schedule\n";
+      
+      //A 3D vector to hold the GIDs so we can see how this process will receive
+      //the vertices in each round, from each process. This way we can reorder things
+      //locally so that the data we receive will automatically construct the CSR correctly
+      //without any other computation. We do the reordering before we start receiving.
+      std::vector<std::vector<std::vector<gno_t>>> recv_GID_per_proc_per_round(
+		      max_rounds+1,std::vector<std::vector<gno_t>>(
+			      comm->getSize(),std::vector<gno_t>(0)));
+      for(int i = 0; i < max_rounds; i++){
+        for(int j = 0; j < comm->getSize(); j++){
+	  recv_GID_per_proc_per_round[i][j] = std::vector<gno_t>(sendcounts[j],0);
+	}
+      }
+
+      if(verbose) std::cout<<comm->getRank()<<": filling out recv GID schedule\n";
+      for(int i = 0; i < comm->getSize(); i++){
+        int curr_round = 0;
+        size_t curr_idx = 0;
+        for(int j = sdispls[i]; j < sdispls[i+1]; j++){
+          if(curr_idx > per_proc_round_vtx_sums[curr_round][i]){
+            curr_round++;
+            curr_idx = 0;
+          }
+          recv_GID_per_proc_per_round[curr_round][i][curr_idx++] = j;
+        }
+      }
+      
+      if(verbose) std::cout<<comm->getRank()<<": reordering gids and degrees in the order they'll be received\n";
+      //reorder the GIDs and degrees locally so that the data received on this process
+      //just happens to be the correct CSR.
+      std::vector<gno_t> final_gid_vec(sendcount, 0);
+      std::vector<offset_t> final_degree_vec(sendcount,0);
+      gno_t reorder_idx = 0;
+      for(int i = 0; i < max_rounds; i++){
+        for(int j = 0; j < comm->getSize(); j++){
+          for(size_t k = 0; k < per_proc_round_vtx_sums[i][j]; k++){
+            final_gid_vec[reorder_idx] = sendbuf[recv_GID_per_proc_per_round[i][j][k]];
+            final_degree_vec[reorder_idx++] = recvDegrees[recv_GID_per_proc_per_round[i][j][k]];
+          }
+        }
+      }
+      //do a quick check to see if we ended up reorganizing anything
+      bool reorganized = false;
+      for(int i = 0; i < sendcount; i++){
+        if(final_gid_vec[i] != sendbuf[i]) reorganized = true;
+      }
+
+      //if we have more than a single round of communication, we need to reorganize.
+      //this alerts of unnecessary reorganization, and a failure to perform necessary
+      //reorganization.
+      if(verbose){
+        if(!reorganized && (max_rounds > 1)) std::cout<<comm->getRank()<<": did not reorgainze GIDs, but probably should have\n";
```

Comment:
It was used for debugging, but i think it would be useful to keep around. I've surrounded the entire test with a verbose check, so we only compute reorganized if verbose is true.